### ISSUE TITLE
shadow-piercing descendant combinator (/deep/) not supported

### ISSUE THREAD

**Issue by binarious** [stylelint/stylelint#2506]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

The [shadow-piercing descendant combinator](https://drafts.csswg.org/css-scoping/#deep-combinator) (`/deep/` and `>>>`) isn't supported. 

> What CSS is needed to reproduce this issue?

```css
.foo /deep/ .bar {
  color: pink;
}
```

> What stylelint configuration is needed to reproduce this issue?

```json
{
  "extends": "stylelint-config-standard"
}
```

> Which version of stylelint are you using?

`7.10.1`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI with `stylelint "src/**/*.{scss,css,styl}"`

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

It's related to the draft [CSS Scoping Module Level 1](https://drafts.csswg.org/css-scoping/) which is already used in [Angular](https://angular.io/docs/ts/latest/guide/component-styles.html#!#-deep-).

> What did you expect to happen?

No warnings to be flagged.

> What actually happened (e.g. what warnings or errors you are getting)?

The following warnings were flagged using `/deep/`:
```shell
src/styles.scss
 18:7  ✖  Unexpected unknown type selector "/deep/"   selector-type-no-unknown
```

The following warnings were flagged using `>>>`:
```shell
src/styles.scss
 18:7  ✖  Expected single space after ">"    selector-combinator-space-after
 18:7  ✖  Unexpected ">>>"                   selector-descendant-combinator-no-non-space
 18:8  ✖  Expected single space after ">"    selector-combinator-space-after
 18:8  ✖  Expected single space before ">"   selector-combinator-space-before
 18:9  ✖  Expected single space before ">"   selector-combinator-space-before
```

**Comment by evilebottnawi** [comment]
@binarious can you use more bigger css it is allow to understand the whole set of rules that do not handle it :smile:

**Comment by evilebottnawi** [comment]
@binarious also i don't see `/deep/` in spec, can you give me link on spec this feature?

**Comment by evilebottnawi** [comment]
@stylelint/core Someone is well versed in angulare(their css style and other stuff) be good add system tests

**Comment by evilebottnawi** [comment]
Also with regard to me, in our team began to appear projects on Angulyar, now I looked and saw that some guys just turn these rules off, I think we should get support :trophy:

**Comment by binarious** [comment]
@evilebottnawi 
> can you use more bigger css

What do you mean by that? The only thing that's required to get stylelint to fail is the use of `/deep/` or `>>>` which is between two selectors.

> also i don't see /deep/ in spec, can you give me link on spec this feature?

I think it's still a draft: https://drafts.csswg.org/css-scoping/#deep-combinator and
https://www.w3.org/TR/css-scoping-1/#deep-combinator
Maybe similar: https://www.w3.org/TR/2013/WD-selectors4-20130502/#idref-combinators
Some discussion about this at the repo of vscode: https://github.com/Microsoft/vscode/issues/7002

**Comment by evilebottnawi** [comment]
@binarious Initial problems are solved, only `selector-combinator-space-before/after` don't report on `/deep/` (and other reference combinators selectors), but i'll figure out how to solve it therefore issue not close :rocket:

**Comment by abdonrd** [comment]
@evilebottnawi you can see the Shadow-Piercing descendant combinator here: https://www.chromestatus.com/features/6750456638341120

But it is deprecated.

**Comment by evilebottnawi** [comment]
@abdonrd thanks for information,
/cc @crisbeto @jelbourn @DevVersion @kara @mmalerba
I apologize if someone missed itю


Does it make sense to introduce support `/deep/` (maybe other reference combinators), it looks like as
deprecated?

**Comment by binarious** [comment]
@abdonrd That's right, it seems deprecated. Angular emulates `/deep/` though.

**Comment by davidtheclark** [comment]
Closed by #2508.

**Comment by jeddy3** [comment]
> But it is deprecated.

@abdonrd Thanks for the heads up!

> Initial problems are solved, only `selector-combinator-space-before/after` don't report on `/deep/`

As the Shadow-Piercing descendant combinator is deprecated I don't think it's prudent to _add code in the next major_ to enforce a specific spacing around it.

I think the work in #2508 and #2509 is still valid though, as they _fix_ false positives associated with this, albeit deprecated, feature. These are minor non-breaking changes that help users.

> Closed by #2508.

👍

**Comment by Mouvedia** [comment]
> [@vue/compiler-sfc] ::v-deep usage as a combinator has been deprecated. Use :deep(<inner-selector>) instead.

@jeddy3 should selector-pseudo-class-no-unknown ignore `:deep()`?

**Comment by jeddy3** [comment]
> @jeddy3 should selector-pseudo-class-no-unknown ignore :deep()?

@Mouvedia If it's in the specs, then yes. If it's a Vue specific thing, then no.

If it's the former, then a pull request would be very welcome.
